data/tutorials/language/0lg_10_polymorphic_variants.md

```diff
@@ -0,0 +1,713 @@
+-->
+
+### A Note on Simple Variants and Polymorphism
+
+The type expression `'a list` does not designate a single type; it designates a family of types, basically all the types that can be created by substituting an actual type to type parameter `'a`. The type expressions `int list`, `bool option list`, or `(float -> float) list list` are real types. They're actual members of the _type family_ `'a list`. Types are intended to have inhabitants, type families don't.
+
+The identifiers `list`, `option`, and others are _type operators_. Just like functions, they take parameters. Although these parameters are not values, they are types. Their results aren't values but types, too.
+
+Simple variants are polymorphic, but not in the same sense as polymorphic variants.
+- Simple variants have [parametric polymorphism](https://en.wikipedia.org/wiki/Parametric_polymorphism)
+- Polymorphic variants have a form of [structural polymorphism](https://en.wikipedia.org/wiki/Structural_type_system)
+
+## A First Example
+
+### Creating Polymorphic Variant Types From Pattern Matching
+
+The visual signature of the polymorphic variants is the back quote. Pattern matching on them looks just the same as with simple variants, except for back quotes (and capitals, which are no longer required).
+```ocaml
+# let f = function `Broccoli -> "Broccoli" | `Fruit name -> name;;
+val f : [< `Broccoli | `Fruit of string ] -> string = <fun>
+```
+
+Here`` `Broccoli`` and`` `Fruit`` play a role similar to the one played by the constructors `Broccoli` and `Fruit` in a variant declared as `type t = Broccoli | Fruit of string`. Except, and most importantly, that the definition doesn't need to be written. The tokens`` `Broccoli`` and `` `Fruit`` are called _tags_ instead of constructors. A tag is defined by a name and a list of parameter types.
+
+The expression ``[< `Broccoli | `Fruit of string ]`` plays the role of a type. However, it does not represent a single type, it represents three different types.
+- The type that only has `` `Broccoli`` as an inhabitant, its translation into a simple variant is `type t0 = Broccoli`
+- The type that only has `` `Fruit`` as an inhabitant, its translation into a simple variant is `type t1 = Fruit of string`
+- Type type that has both `` `Broccoli`` and `` `Fruit`` inhabitants, its translation into a simple variant is `type t2 = Broccoli | Fruit of string`
+
+Note each of the above translations into simple variants is correct. However, entering them as-is into the environment would lead to constructor shadowing (unless type annotation is used at pattern or expression level, see section on [Shared Constructors](#Shared-Constructors)).
+
+This also illustrates the other striking feature of polymorphic variants: values can be attached to several types. For instance, the tag `` `Broccoli`` inhabits ``[ `Broccoli ]`` and``[ `Broccoli | `Fruit of String ]``, but it also inhabits any type that contains it.
```

Comment:
> "but it also inhabits any type that contains it."

This is going to raise eyebrows. Inhabits alone is a term that confused the hell out of me for a long time.

I think if we build the intuition that `` `Broccoli `` is a value that belongs to the type `` [ > `Broccoli ] `` then it would make more sense that anywhere that we see `` `Broccoli `` in the type, then `` `Broccoli `` is a valid value.

Of course there's the case where `` `Broccoli `` belongs to `` [ > `Apple ] `` because of `>` but I'd leave `>` until we talk about polyvars as function inputs/outputs.

Reply:
Agreed. The terminology I've used here is clumsy. This is needs to be rewritten.